Bootcamp2015-AmazingRace/Views/MapPage.xaml.cs

```diff
@@ -0,0 +1,165 @@
+            {
+                UpdateMyLocation(args.Position);
+            });
+        }
+
+        public ObservableCollection<PinViewModel> Pins { get; set; }
+
+
+        private async void OnRefreshClick(object sender, RoutedEventArgs e)
+        {
+            await Refresh();
+        }
+
+        private async void OnAddPinClick(object sender, RoutedEventArgs e)
+        {
+            var pin = new PinViewModel
+            {
+                Text = "point",
+                Color = new SolidColorBrush(Colors.Red),
+                Location = _map.Center,
+                CanGetDirections = true,
+            };
+
+            Pins.Add(pin);
+        }
+
+        private async Task Refresh()
+        {
+            var pos = await _locator.GetGeopositionAsync();
+            UpdateMyLocation(pos);
+        }
+
+        private void UpdateMyLocation(Geoposition p)
+        {
+            var pos = new Geopoint(p.Coordinate.Point.Position);
+            _map.Center = pos;
+
+            var pin = new PinViewModel
+            {
+                Text = "current location",
+                Color = new SolidColorBrush(Colors.Green),
+                Location = pos
+            };
+
+            if (Pins.Count > 0)
+            {
+                Pins.RemoveAt(0);
+            }
+            Pins.Insert(0, pin);
+        }
+
+        #region Directions
+
+        private void OnDirectionsClick(object sender, RoutedEventArgs e)
+        {
+            var pin =  ((sender as FrameworkElement).DataContext) as PinViewModel;
+            ShowDirectionsExternal(pin);
+        }
+
+        private async void ShowDirectionsInApp(PinViewModel pin)
+        {
+            var routeResult = await MapRouteFinder.GetDrivingRouteAsync(
+                            Pins[0].Location,
+                            pin.Location,
+                            MapRouteOptimization.Time,
+                            MapRouteRestrictions.None);
+
+            var viewOfRoute = new MapRouteView(routeResult.Route);
+            viewOfRoute.RouteColor = Colors.Yellow;
+            viewOfRoute.OutlineColor = Colors.Black;
+            _map.Routes.Clear();
+            _map.Routes.Add(viewOfRoute);
+            await _map.TrySetViewBoundsAsync(routeResult.Route.BoundingBox, new Thickness(10), MapAnimationKind.Bow);
+        }
+
+        private async void ShowDirectionsExternal(PinViewModel pin)
+        {
+            var latitude = pin.Location.Position.Latitude;
+            var longitude = pin.Location.Position.Longitude;
+            string name = "My next clue";
+
+            // Assemble the Uri to launch.
+            Uri uri = new Uri("ms-drive-to:?destination.latitude=" + latitude + "&destination.longitude=" + longitude + "&destination.name=" + name);
+
+            // Launch the Uri.
+            var success = await Windows.System.Launcher.LaunchUriAsync(uri);
+        }
+
+        #endregion
+
+    }
+
+    public class PinViewModel
```

Comment:
ViewModels should implement INotifyPropertyChanged. Caliburn provides a base class BasePropertyChanged for this. As it stands, changes to these viewmodels won't be updated in the UI.